src/coreclr/interpreter/compiler.cpp

```diff
@@ -5161,12 +5181,16 @@ void InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
             }
 
             case CEE_ISINST:
+            case CEE_CASTCLASS:
             {
                 CHECK_STACK(1);
                 CORINFO_RESOLVED_TOKEN resolvedToken;
                 ResolveToken(getU4LittleEndian(m_ip + 1), CORINFO_TOKENKIND_Casting, &resolvedToken);
 
-                CorInfoHelpFunc castingHelper = m_compHnd->getCastingHelper(&resolvedToken, false /* throwing */);
+                bool isCastClass = (opcode == CEE_CASTCLASS);
+
+                CorInfoHelpFunc castingHelper = m_compHnd->getCastingHelper(&resolvedToken, isCastClass);
+
                 AddIns(INTOP_CALL_HELPER_PP_2);
                 m_pLastNewIns->data[0] = GetDataItemIndexForHelperFtn(castingHelper);
                 m_pLastNewIns->data[1] = GetDataItemIndex(resolvedToken.hClass);
```

Comment:
`resolvedToken.hClass` won't work for shared generics here. The direction you took with EmitGenericHandle was good.